File: blockchain.py
import hashlib
import json
import string
import time


class Block:
    def __init__(self, index: int, previous_hash: str):
        self.index = index
        self.timestamp = time.time_ns() // 1_000_000
        self.previous_hash = previous_hash
        self.data: list[dict] = []
        self.nonce = ""
        self.hash = self.calculate_hash()

    def calculate_hash(self) -> str:
        payload = json.dumps(
            [self.index, self.timestamp, self.previous_hash, self.data, self.nonce],
            sort_keys=True,
        ).encode("utf-8")
        return hashlib.sha256(payload).hexdigest()

    def mine_block(self, difficulty: int) -> None:
        chars = string.ascii_letters + string.digits
        target = "0" * difficulty
        nonce = 0

        while True:
            self.nonce = f"{chars[nonce % len(chars)]}{nonce}"
            self.hash = self.calculate_hash()
            if self.hash.startswith(target):
                return
            nonce += 1

    def add_bet(self, prediction_id: str, user_id: str, amount: float, is_yes: bool) -> None:
        self.data.append(
            {
                "prediction_id": prediction_id,
                "user_id": user_id,
                "amount": amount,
                "bet_type": "yes" if is_yes else "no",
            }
        )


class Blockchain:
    def __init__(self):
        genesis = self.create_genesis_block()
        self.chain = [genesis]
        self.current = Block(1, genesis.hash)
        self.difficulty = 4

    def create_genesis_block(self) -> Block:
        return Block(0, "0")

    def get_latest_block(self) -> Block:
        return self.chain[-1]

    def add_block(self, new_block: Block) -> None:
        new_block.previous_hash = self.get_latest_block().hash
        new_block.hash = new_block.calculate_hash()
        self.chain.append(new_block)

    def is_chain_valid(self) -> bool:
        for index in range(1, len(self.chain)):
            current_block = self.chain[index]
            previous_block = self.chain[index - 1]
            if current_block.hash != current_block.calculate_hash():
                return False
            if current_block.previous_hash != previous_block.hash:
                return False
        return True

    def delete_block(self, index: int) -> None:
        del self.chain[index]
        for position in range(max(index, 1), len(self.chain)):
            self.chain[position].previous_hash = self.chain[position - 1].hash
            self.chain[position].hash = self.chain[position].calculate_hash()

    def mine_current(self) -> None:
        self.current.mine_block(self.difficulty)
        self.add_block(self.current)
        self.current = Block(self.current.index + 1, self.get_latest_block().hash)
lf.current = block(self.current.index + 1, previousHash=self.getLatestBlock().hash)

    

File: static/blockchain.js
export class Block {
    constructor(index, previousHash) {
        this.index = index;
        this.timestamp = Date.now();
        this.previousHash = previousHash;
        this.data = [];
        this.nonce = "";
        this.hash = this.calculateHash();
    }

    calculateHash() {
        const contents = JSON.stringify([
            this.index,
            this.timestamp,
            this.previousHash,
            this.data,
            this.nonce,
        ]);
        return stableHexHash(contents);
    }

    mineBlock(difficulty) {
        const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".split("");
        const target = "0".repeat(difficulty);

        for (let len = 1; ; len++) {
            for (const combo of combinations(chars, len)) {
                this.nonce = combo.join("");
                this.hash = this.calculateHash();
                if (this.hash.startsWith(target)) {
                    return;
                }
            }
        }
    }

    fakeMine() {
        this.nonce = "fake!";
        this.hash = this.calculateHash();
    }

    addTransaction(transaction) {
        this.data.push({
            ...transaction,
        });
    }
}

export class Blockchain {
    constructor() {
        this.difficulty = 4;
        this.queue = new TransactionQueue();
        const genesis = this.createGenesisBlock();
        this.current = new Block(1, genesis.hash);
        this.chain = [genesis, this.current];

        this.queue.onChange = () => {
            if (this.queue.size() >= 200) {
                this.sealBlock();
            }
        };
    }

    export(){
        return {
            difficulty: this.difficulty,
            chain: this.chain, 
            current: this.current,
            queue: this.queue.transactions,
        }
    }

    import(data) {
        this.difficulty = data.difficulty;
        this.chain = data.chain;
        this.current = data.current;
        this.queue.transactions = data.queue;
    }

    createGenesisBlock() {
        return new Block(0, "0");
    }

    getLatestBlock() {
        return this.chain[this.chain.length - 1];
    }

    addBlock(newBlock) {
        newBlock.previousHash = this.getLatestBlock().hash;
        newBlock.hash = newBlock.calculateHash();
        this.chain.push(newBlock);
    }

    sealBlock() {
        const transactions = this.queue.drain(100);
        for (const tx of transactions) {
            this.current.addTransaction(tx);
        }

        this.current.fakeMine();
        this.addBlock(this.current);
        this.current = new Block(this.current.index + 1, this.getLatestBlock().hash);
    }

    enqueueEvent(eventType, payload = {}) {
        this.queue.push({
            event_type: eventType,
            ...payload,
        });
    }

    isChainValid() {
        for (let i = 1; i < this.chain.length; i++) {
            const current = this.chain[i];
            const previous = this.chain[i - 1];
            if (current.hash !== current.calculateHash()) {
                return false;
            }
            if (current.previousHash !== previous.hash) {
                return false;
            }
        }

        return true;
    }

    deleteBlock(index) {
        this.chain.splice(index, 1);
        for (let i = Math.max(index, 1); i < this.chain.length; i++) {
            this.chain[i].previousHash = this.chain[i - 1].hash;
            this.chain[i].hash = this.chain[i].calculateHash();
        }
    }

    mineCurrent() {
        this.current.mineBlock(this.difficulty);
        this.addBlock(this.current);
        this.current = new Block(this.current.index + 1, this.getLatestBlock().hash);
    }
}

export class TransactionQueue {
    constructor() {
        this.transactions = [];
        this.onChange = null;
    }

    push(tx) {
        tx._hash = hashCode(JSON.stringify(tx));
        this.transactions.push(tx);
        this.transactions.sort((a, b) => {
            if (a.timestamp !== b.timestamp) {
                return a.timestamp - b.timestamp;
            }
            return a._hash - b._hash;
        });
        this.onChange?.();
    }

    drain(n) {
        return this.transactions.splice(0, n);
    }

    size() {
        return this.transactions.length;
    }

    peek() {
        return this.transactions[0] ?? null;
    }
}

function hashCode(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        hash = ((hash << 5) - hash) + str.charCodeAt(i);
        hash |= 0;
    }
    return hash;
}

function stableHexHash(str) {
    const seedA = hashCode(str);
    const seedB = hashCode(`${str}|b`);
    const seedC = hashCode(`${str}|c`);
    const seedD = hashCode(`${str}|d`);

    return [seedA, seedB, seedC, seedD]
        .map((part) => (part >>> 0).toString(16).padStart(8, "0"))
        .join("");
}

function* combinations(arr, len) {
    if (len === 1) {
        for (const item of arr) {
            yield [item];
        }
        return;
    }

    for (let i = 0; i <= arr.length - len; i++) {
        for (const rest of combinations(arr.slice(i + 1), len - 1)) {
            yield [arr[i], ...rest];
        }
    }
}